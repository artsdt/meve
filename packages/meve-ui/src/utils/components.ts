import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>;
  unmount(): void;
}

export function mountComponent(component: ComponentOptions<Vue>, container: string, options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}

export function nextTick(): Promise<void> {
  return new Promise((resolve) => Vue.nextTick(resolve))
}
