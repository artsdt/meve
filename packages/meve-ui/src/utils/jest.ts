import Vue from 'vue'
import { Wrapper, config } from '@vue/test-utils'
import { isPlainObject } from './shared'

export const delay = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export function mockOffset({
  offsetWidth,
  offsetHeight,
  clientWidth,
  clientHeight,
  offsetLeft,
  offsetTop,
  scrollWidth,
  scrollHeight,
}: {
  offsetWidth?: number;
  offsetHeight?: number;
  clientWidth?: number;
  clientHeight?: number;
  offsetLeft?: number;
  offsetTop?: number;
  scrollWidth?: number;
  scrollHeight?: number;
} = {}) {
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return offsetWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    offsetHeight: {
      get() {
        return offsetHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    clientWidth: {
      get() {
        return clientWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    clientHeight: {
      get() {
        return clientHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    offsetLeft: {
      get() {
        return offsetLeft ?? (parseFloat(window.getComputedStyle(this).marginLeft) || 0)
      },
    },
    offsetTop: {
      get() {
        return offsetTop ?? (parseFloat(window.getComputedStyle(this).marginTop) || 0)
      },
    },
    offsetParent: {
      get() {
        return this.parentNode ?? {}
      },
    },
    scrollWidth: {
      get() {
        return scrollWidth ?? 0
      },
    },
    scrollHeight: {
      get() {
        return scrollHeight ?? 0
      },
    },
  })

  return {
    mockRestore() {
      Object.defineProperties(HTMLElement.prototype, {
        offsetWidth: {
          get() {
            return parseFloat(window.getComputedStyle(this).width) || 0
          },
        },
        offsetHeight: {
          get() {
            return parseFloat(window.getComputedStyle(this).height) || 0
          },
        },
        clientWidth: {
          get() {
            return parseFloat(window.getComputedStyle(this).width) || 0
          },
        },
        clientHeight: {
          get() {
            return parseFloat(window.getComputedStyle(this).height) || 0
          },
        },
        offsetLeft: {
          get() {
            return parseFloat(window.getComputedStyle(this).marginLeft) || 0
          },
        },
        offsetTop: {
          get() {
            return parseFloat(window.getComputedStyle(this).marginTop) || 0
          },
        },
        offsetParent: {
          get() {
            return this.parentNode ?? {}
          },
        },
        scrollWidth: {
          get() {
            return 0
          },
        },
        scrollHeight: {
          get() {
            return 0
          },
        },
      })
    },
  }
}

export async function triggerDrag(
  el: Wrapper<Vue> | HTMLElement | Window,
  x = 0,
  y = 0,
  triggerTouchend = true,
  triggerDocument = false
) {
  const delegationEl = triggerDocument ? document : el

  await trigger(el, 'touchstart', 0, 0)
  await trigger(delegationEl, 'touchmove', x / 4, y / 4)
  await trigger(delegationEl, 'touchmove', x / 3, y / 3)
  await trigger(delegationEl, 'touchmove', x / 2, y / 2)
  await trigger(delegationEl, 'touchmove', x, y)

  if (triggerTouchend) {
    await trigger(delegationEl, 'touchend', x, y)
  }
}

export function getTouch(el: Element | Window | Document, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: Wrapper<Vue> | HTMLElement | Window | Document,
  eventName: string,
  x = 0,
  y = 0,
  deltaX = 0,
  deltaY = 0,
  offsetX = 0,
  offsetY = 0,
  args = {}
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    deltaX,
    deltaY,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
    ...args,
  })

  el.dispatchEvent(event)

  return Vue.nextTick()
}

export function mockStubs() {
  const originStubs = config.stubs

  config.stubs = {}

  return {
    mockRestore() {
      config.stubs = originStubs
    },
  }
}

export function mockDoubleRaf() {
  const originMethod = window.requestAnimationFrame

  Object.assign(window, {
    requestAnimationFrame(fn: any) {
      setTimeout(fn, 16)
    },
  })

  return {
    mockRestore() {
      window.requestAnimationFrame = originMethod
    },
  }
}

export function mockScrollTo(Element: any) {
  Element.prototype.scrollTo = function (x: ScrollToOptions | number, y?: number) {
    if (isPlainObject(x)) {
      this.scrollLeft = x.left
      this.scrollTop = x.top
    } else {
      this.scrollLeft = x
      this.scrollTop = y
    }
  }
}
