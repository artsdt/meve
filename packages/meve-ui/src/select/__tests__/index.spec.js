import Select from '..'
import Option from '../../option'
import { mount } from '@vue/test-utils'
import { delay, mockStubs, trigger } from '../../utils/jest'

async function clickOpen(wrapper) {
  wrapper.find('.m-select__select').trigger('click')
  await delay(300)
}

function expectPopoverShow() {
  expect(document.querySelector('.m-popover__container').style.display).toBe('')
}

function expectPopoverHide() {
  expect(document.querySelector('.m-popover__container').style.display).toBe('none')
}

async function tabFocus(wrapper) {
  await trigger(wrapper.find('.m-select__input'), 'focus')
}

async function mouseenterOption() {
  await trigger(document.querySelector('.m-option'), 'mouseenter')
}

async function selectOption() {
  await trigger(document.querySelector('.m-option'), 'click')
  await delay(300)
}

async function selectAllOption() {
  for (const option of document.querySelectorAll('.m-option')) {
    await trigger(option, 'touchstart')
    await trigger(option, 'click')
  }
  await delay(300)
}

async function selectAllOptionByKeyboard() {
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'ArrowDown' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'ArrowDown' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
}

async function unSelectAllOptionByKeyboard() {
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'ArrowUp' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'ArrowUp' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
}

async function selectOptionByKeyboard() {
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'ArrowDown' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
  await delay(300)
}

async function closeOption(wrapper) {
  const tag = wrapper.find('.m-tag__close-button')

  await trigger(tag, 'click')
  await delay(100)
}

function expectClearIconShow(wrapper) {
  expect(wrapper.find('.m-select__clear-icon')).toBeTruthy()
  expect(wrapper.find('.m-select--clear-icon-visible')).toBeTruthy()
}

function expectClearIconHide(wrapper) {
  expect(wrapper.find('.m-select__clear-icon')).toBeTruthy()
  expect(wrapper.find('.m-select__clear-icon').classes().includes('.m-select--clear-icon-visible')).toBeFalsy()
}

async function clickClearIcon(wrapper) {
  await wrapper.find('.m-select__clear-icon').trigger('click')
}

function destroy(wrapper) {
  wrapper.destroy()
  document.body.innerHTML = ''
}

test('test select option', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await clickOpen(wrapper)
  expect(wrapper.emitted().click.length).toBe(1)
  expect(wrapper.emitted().focus.length).toBe(1)
  expectPopoverShow()

  await selectOption()

  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toBe('Eat')
  expect(wrapper.emitted().select.length).toBe(1)
  expect(wrapper.emitted().select[0][0]).toBe('Eat')
  expect(wrapper.emitted().blur.length).toBe(1)
  expectPopoverHide()

  destroy(wrapper)
})

test('test select option use keyboard', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await clickOpen(wrapper)
  expect(wrapper.emitted().click.length).toBe(1)
  expect(wrapper.emitted().focus.length).toBe(1)
  expectPopoverShow()

  await selectOptionByKeyboard(wrapper)

  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toBe('Sleep')
  expect(wrapper.emitted().select.length).toBe(1)
  expect(wrapper.emitted().select[0][0]).toBe('Sleep')
  expect(wrapper.emitted().blur.length).toBe(1)
  expectPopoverHide()

  destroy(wrapper)
})

test('test select option relation value', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" :value="1" />
        <m-option label="Sleep" :value="2" />
        <m-option label="Coding" :value="3"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await clickOpen(wrapper)
  expect(wrapper.emitted().click.length).toBe(1)
  expect(wrapper.emitted().focus.length).toBe(1)
  expectPopoverShow()

  await selectOption()

  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toBe(1)
  expect(wrapper.emitted().select.length).toBe(1)
  expect(wrapper.emitted().select[0][0]).toBe(1)
  expect(wrapper.emitted().blur.length).toBe(1)
  expectPopoverHide()

  destroy(wrapper)
})

test('test select size', async () => {
  const wrapper = mount(
    {
      components: {
        [Select.name]: Select,
        [Option.name]: Option,
      },
      data: () => ({
        value: '',
      }),
      template: `
      <div>
        <m-select v-model="value">
          <m-option label="Eat"/>
        </m-select>
        <m-select size="small" v-model="value">
          <m-option label="Eat"/>
        </m-select>
        <m-select size="mini" v-model="value">
          <m-option label="Eat"/>
        </m-select>
      </div>
    `,
    },
    { attachTo: document.body }
  )

  expect(wrapper.html()).toMatchSnapshot()
  destroy(wrapper)
})

test('test select option clear', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: 'Eat', clearable: true },
      listeners: {
        input(v) {
          wrapper.setProps({ value: v })
        },
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  expectClearIconShow(wrapper)
  await clickClearIcon(wrapper)

  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toBe(undefined)
  expect(wrapper.emitted().clear.length).toBe(1)
  expect(wrapper.emitted().clear[0][0]).toBe(undefined)
  expectClearIconHide(wrapper)

  destroy(wrapper)
})

test('test toggle option in multiple', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: {
        value: [],
        multiple: true,
      },
      listeners: {
        input(v) {
          wrapper.setProps({ value: v })
        },
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )
  await clickOpen(wrapper)
  expect(wrapper.emitted().click.length).toBe(1)
  expect(wrapper.emitted().focus.length).toBe(1)
  expectPopoverShow()

  await selectAllOption()
  // cancel
  await selectAllOption()

  expect(wrapper.emitted().input.length).toBe(6)
  expect(wrapper.emitted().input[0][0]).toStrictEqual(['Eat'])
  expect(wrapper.emitted().input[1][0]).toStrictEqual(['Eat', 'Sleep'])
  expect(wrapper.emitted().input[2][0]).toStrictEqual(['Eat', 'Sleep', 'Coding'])
  expect(wrapper.emitted().input[3][0]).toStrictEqual(['Sleep', 'Coding'])
  expect(wrapper.emitted().input[4][0]).toStrictEqual(['Coding'])
  expect(wrapper.emitted().input[5][0]).toStrictEqual([])

  destroy(wrapper)
})

test('test toggle option use keyboard in multiple', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: {
        value: [],
        multiple: true,
      },
      listeners: {
        input(v) {
          wrapper.setProps({ value: v })
        },
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )
  await clickOpen(wrapper)
  expect(wrapper.emitted().click.length).toBe(1)
  expect(wrapper.emitted().focus.length).toBe(1)
  expectPopoverShow()

  await selectAllOptionByKeyboard()
  await unSelectAllOptionByKeyboard()

  expect(wrapper.emitted().input.length).toBe(6)
  expect(wrapper.emitted().input[0][0]).toStrictEqual(['Eat'])
  expect(wrapper.emitted().input[1][0]).toStrictEqual(['Eat', 'Sleep'])
  expect(wrapper.emitted().input[2][0]).toStrictEqual(['Eat', 'Sleep', 'Coding'])
  expect(wrapper.emitted().input[3][0]).toStrictEqual(['Eat', 'Sleep'])
  expect(wrapper.emitted().input[4][0]).toStrictEqual(['Eat'])
  expect(wrapper.emitted().input[5][0]).toStrictEqual([])

  destroy(wrapper)
})

test('test select option clear in multiple', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: {
        value: ['Eat', 'Sleep', 'Coding'],
        multiple: true,
        clearable: true,
      },
      listeners: {
        input(v) {
          wrapper.setProps({ value: v })
        },
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )
  expectClearIconShow(wrapper)
  await clickClearIcon(wrapper)

  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toStrictEqual([])
  expect(wrapper.emitted().clear.length).toBe(1)
  expect(wrapper.emitted().clear[0][0]).toStrictEqual([])
  expectClearIconHide(wrapper)

  destroy(wrapper)
})

test('test select disabled', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: {
        value: 'Eat',
        disabled: true,
        clearable: true,
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await clickOpen(wrapper)
  expectPopoverHide()
  await tabFocus(wrapper)
  expectPopoverHide()

  expectClearIconShow(wrapper)
  await clickClearIcon(wrapper)

  expect(wrapper.emitted().input).toBeFalsy()
  expect(wrapper.emitted().clear).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()

  destroy(wrapper)
})

test('test select readonly', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: {
        value: 'Eat',
        readonly: true,
        clearable: true,
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await clickOpen(wrapper)
  expectPopoverHide()
  await tabFocus(wrapper)
  expectPopoverHide()

  expectClearIconShow(wrapper)
  await clickClearIcon(wrapper)

  expect(wrapper.emitted().input).toBeFalsy()
  expect(wrapper.emitted().clear).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()

  destroy(wrapper)
})

test('test select validation', async () => {
  const { mockRestore } = mockStubs()

  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: {
        value: '',
        rules: [(v) => !!v || '不能为空'],
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  wrapper.vm.validate()
  await delay(100)
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.resetValidation()
  await delay(100)
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.reset()
  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toBe(undefined)
  mockRestore()

  destroy(wrapper)
})

test('test select validation in multiple', async () => {
  const { mockRestore } = mockStubs()

  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: {
        value: [],
        multiple: true,
        rules: [(v) => v.length > 0 || '不能为空'],
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  wrapper.vm.validate()
  await delay(100)
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.resetValidation()
  await delay(100)
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.reset()
  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toStrictEqual([])
  mockRestore()

  destroy(wrapper)
})

test('test select focus and blur', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  wrapper.vm.focus()
  await delay(300)
  expectPopoverShow()

  wrapper.vm.blur()
  await delay(300)
  expectPopoverHide()

  expect(wrapper.emitted().focus.length).toBe(1)
  expect(wrapper.emitted().blur.length).toBe(1)

  destroy(wrapper)
})

test('test select close in multiple', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: {
        value: ['Eat', 'Sleep', 'Coding'],
        multiple: true,
      },
      listeners: {
        input(v) {
          wrapper.setProps({ value: v })
        },
      },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await closeOption(wrapper)
  await closeOption(wrapper)
  await closeOption(wrapper)
  expect(wrapper.emitted().input[0][0]).toStrictEqual(['Sleep', 'Coding'])
  expect(wrapper.emitted().input[1][0]).toStrictEqual(['Coding'])
  expect(wrapper.emitted().input[2][0]).toStrictEqual([])
  expect(wrapper.emitted()['tag-close'][0][0]).toStrictEqual(['Sleep', 'Coding'])
  expect(wrapper.emitted()['tag-close'][1][0]).toStrictEqual(['Coding'])
  expect(wrapper.emitted()['tag-close'][2][0]).toStrictEqual([])

  destroy(wrapper)
})

test('test select tab focus and esc blur', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await tabFocus(wrapper)
  await delay(300)
  expectPopoverShow()

  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Escape' })
  await delay(300)
  expectPopoverHide()

  expect(wrapper.emitted().focus.length).toBe(1)
  expect(wrapper.emitted().blur.length).toBe(1)

  destroy(wrapper)
})

test('test select tab focus and touch document blur', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await tabFocus(wrapper)
  await delay(300)
  expectPopoverShow()

  await trigger(document, 'touchstart')
  await delay(300)
  expectPopoverHide()

  expect(wrapper.emitted().focus.length).toBe(1)
  expect(wrapper.emitted().blur.length).toBe(1)

  destroy(wrapper)
})

test('test select popover events', async () => {
  const { mockRestore } = mockStubs()

  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await clickOpen(wrapper)
  expect(wrapper.emitted().open.length).toBe(1)

  await selectOption()
  expect(wrapper.emitted().close.length).toBe(1)
  expect(wrapper.emitted().closed.length).toBe(1)

  mockRestore()
  destroy(wrapper)
})

test('test select slots', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
        'prepend-icon': '<div>prepend icon</div>',
        'append-icon': '<div>append icon</div>',
      },
    },
    { attachTo: document.body }
  )

  expect(wrapper.html()).toMatchSnapshot()

  destroy(wrapper)
})

test('test select keyboard trigger on popover hide', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'ArrowDown' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'ArrowUp' })
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Escape' })

  await delay(300)
  expect(Object.keys(wrapper.emitted()).length).toBe(0)
  destroy(wrapper)
})

test('test select option mouseenter', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" />
        <m-option label="Sleep" />
        <m-option label="Coding"/>
      `,
      },
    },
    { attachTo: document.body }
  )

  await clickOpen(wrapper)
  await mouseenterOption()
  expect(document.querySelector('.m-option').classList.contains('m-option--key-active')).toBeTruthy()

  destroy(wrapper)
})

test('test select option all disabled', async () => {
  const wrapper = mount(
    Select,
    {
      components: {
        [Option.name]: Option,
      },
      propsData: { value: '' },
      slots: {
        default: `
        <m-option label="Eat" disabled />
        <m-option label="Sleep" disabled />
        <m-option label="Coding" disabled />
      `,
      },
    },
    { attachTo: document.body }
  )

  await delay(16)
  await clickOpen(wrapper)
  expectPopoverHide()
  destroy(wrapper)
})
