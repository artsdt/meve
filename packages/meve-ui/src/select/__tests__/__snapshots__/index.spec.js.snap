// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test select disabled 1`] = `
"<div class=\\"m-select m--box\\">
  <div class=\\"m-form-item\\">
    <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
      <div class=\\"m-select__select m-select--normal m-select--disabled\\"><input id=\\"m-select-10\\" autocomplete=\\"off\\" disabled=\\"disabled\\" readonly=\\"true\\" class=\\"m-select__input\\"><i input-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-close-circle m-select__clear-icon m-select--clear-icon-visible\\"></i><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i></div>
      <div class=\\"m-teleport\\"></div>
    </div>
    <transition-stub name=\\"m-form-details\\"></transition-stub>
  </div>
</div>"
`;

exports[`test select readonly 1`] = `
"<div class=\\"m-select m--box\\">
  <div class=\\"m-form-item\\">
    <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
      <div class=\\"m-select__select m-select--normal m-select--outline-hover\\"><input id=\\"m-select-11\\" autocomplete=\\"off\\" readonly=\\"true\\" class=\\"m-select__input\\"><i input-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-close-circle m-select__clear-icon m-select--clear-icon-visible\\"></i><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i></div>
      <div class=\\"m-teleport\\"></div>
    </div>
    <transition-stub name=\\"m-form-details\\"></transition-stub>
  </div>
</div>"
`;

exports[`test select size 1`] = `
"<div>
  <div class=\\"m-select m--box\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
        <div class=\\"m-select__select m-select--normal m-select--outline-hover\\"><input id=\\"m-select-3\\" autocomplete=\\"off\\" readonly=\\"true\\" class=\\"m-select__input\\"><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i></div>
        <div class=\\"m-teleport\\"></div>
      </div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
  </div>
  <div class=\\"m-select m--box\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
        <div class=\\"m-select__select m-select--small m-select--outline-hover\\"><input id=\\"m-select-4\\" autocomplete=\\"off\\" readonly=\\"true\\" class=\\"m-select__input\\"><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i></div>
        <div class=\\"m-teleport\\"></div>
      </div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
  </div>
  <div class=\\"m-select m--box\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
        <div class=\\"m-select__select m-select--mini m-select--outline-hover\\"><input id=\\"m-select-5\\" autocomplete=\\"off\\" readonly=\\"true\\" class=\\"m-select__input\\"><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i></div>
        <div class=\\"m-teleport\\"></div>
      </div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
  </div>
</div>"
`;

exports[`test select slots 1`] = `
"<div class=\\"m-select m--box\\">
  <div class=\\"m-form-item\\">
    <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
      <div class=\\"m-select__select m-select--normal m-select--outline-hover\\">
        <div>prepend icon</div><input id=\\"m-select-19\\" autocomplete=\\"off\\" readonly=\\"true\\" class=\\"m-select__input\\">
        <div>append icon</div><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i>
      </div>
      <div class=\\"m-teleport\\"></div>
    </div>
    <transition-stub name=\\"m-form-details\\"></transition-stub>
  </div>
</div>"
`;

exports[`test select validation 1`] = `
"<div class=\\"m-select m--box\\">
  <div class=\\"m-form-item\\">
    <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
      <div class=\\"m-select__select m-select--normal m-select--outline-hover m-select--validation-error-border\\"><input id=\\"m-select-12\\" autocomplete=\\"off\\" readonly=\\"true\\" class=\\"m-select__input\\"><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i></div>
      <div class=\\"m-teleport\\"></div>
    </div>
    <div class=\\"m-form-details\\">不能为空</div>
  </div>
</div>"
`;

exports[`test select validation 2`] = `
"<div class=\\"m-select m--box\\">
  <div class=\\"m-form-item\\">
    <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
      <div class=\\"m-select__select m-select--normal m-select--outline-hover\\"><input id=\\"m-select-12\\" autocomplete=\\"off\\" readonly=\\"true\\" class=\\"m-select__input\\"><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i></div>
      <div class=\\"m-teleport\\"></div>
    </div>
    <!---->
  </div>
</div>"
`;

exports[`test select validation in multiple 1`] = `
"<div class=\\"m-select m--box\\">
  <div class=\\"m-form-item\\">
    <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
      <div class=\\"m-select__select m-select--normal m-select--outline-hover m-select--validation-error-border\\">
        <div class=\\"m-select__tags\\" style=\\"padding-bottom: 0px;\\"><input readonly=\\"true\\" class=\\"m-select__placeholder\\"></div><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i>
      </div>
      <div class=\\"m-teleport\\"></div>
    </div>
    <div class=\\"m-form-details\\">不能为空</div>
  </div>
</div>"
`;

exports[`test select validation in multiple 2`] = `
"<div class=\\"m-select m--box\\">
  <div class=\\"m-form-item\\">
    <div class=\\"m-popover m-select__popover\\" select-cover=\\"true\\">
      <div class=\\"m-select__select m-select--normal m-select--outline-hover\\">
        <div class=\\"m-select__tags\\" style=\\"padding-bottom: 0px;\\"><input readonly=\\"true\\" class=\\"m-select__placeholder\\"></div><i select-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-menu-down m-select__arrow\\"></i>
      </div>
      <div class=\\"m-teleport\\"></div>
    </div>
    <!---->
  </div>
</div>"
`;
